orttraining/orttraining/models/runner/training_runner.cc

```diff
@@ -634,50 +634,70 @@ void TrainingRunner::RunWithUpdate(VectorString& feed_names,
                                    VectorString& fetch_names,
                                    std::vector<MLValue>& feeds,
                                    std::vector<MLValue>& fetches) {
-  // Cyclically pick up a worker ID.
-  const size_t worker_id = step_ % params_.pipeline_parallel_size;
-
-  // Wait for the previous work to finish its job.
-  // Its resource cannot be overrided when it's still working.
-  pipeline_worker_pool_.Join(worker_id);
-
-  // Copy thread-used variable to thread-specific buffer to maintain their life.
-  pipeline_worker_pool_.worker_states[worker_id].feed_names = feed_names;
-  pipeline_worker_pool_.worker_states[worker_id].feeds = feeds;
-  pipeline_worker_pool_.worker_states[worker_id].fetch_names = fetch_names;
-  pipeline_worker_pool_.worker_states[worker_id].fetches = std::vector<MLValue>();
-
-  Status status = Status::OK();
-  pipeline_worker_pool_.workers[worker_id] = std::thread([&](
-      const size_t worker_id, const size_t step) {
+  if (params_.pipeline_parallel_size > 1) {
+    // Cyclically pick up a worker ID.
+    const size_t worker_id = step_ % params_.pipeline_parallel_size;
+
+    // Wait for the previous work to finish its job.
+    // Its resource cannot be overrided when it's still working.
+    pipeline_worker_pool_.Join(worker_id);
```

Comment:
>pipeline_worker_pool_.Join(worker_id); [](start = 4, length = 38)

The thread should already finish its previous job by the call to JoinAll() in line 675 from previous iteration. So maybe instead of calling Join() again, we can add an assert ensure joinable() is false?

Reply:
I don't understand. We are not sure if the selected thread already dead or still alive.

---
In reply to: [441866766](https://github.com/microsoft/onnxruntime/pull/4265#discussion_r441866766) [](ancestors = 441866766)

Reply:
Calling Join() here is to wait for the previous work from previous call to RunWithUpdate(). But during previous RunWithUpdate(), the job is finished by calling JoinAll() in line 675. So it is guaranteed from the beginning of new RunWithUpdate(), all worker threads are done. Maybe I missed something and there is more work this thread is assigned with?

---
In reply to: [441871710](https://github.com/microsoft/onnxruntime/pull/4265#discussion_r441871710) [](ancestors = 441871710,441866766)

Reply:
It's untrue --- update step is packed with accumulation steps in a pipeline. We here wait for an accumulation thread to finish its job and use that thread index here.

---
In reply to: [441874230](https://github.com/microsoft/onnxruntime/pull/4265#discussion_r441874230) [](ancestors = 441874230,441871710,441866766)